### PR TITLE
Fix glfw usage of callUserCallback

### PR DESCRIPTION
**Pull request by sbc100**
Supersedes #17438 and #17376.

### CODE REVIEW

File: ChangeLog.md [hunk 1]
```diff
@@ -18,8 +18,11 @@ to browse the changes between the tags.
 
 See docs/process.md for more on how version tagging works.
 
-3.1.16
+3.1.17
 ------
+
+3.1.16 - 07/14/2022
+-------------------
```
Comment (kripken) [inline ChangeLog.md:25]: This diff shouldn't appear here?

File: src/Fetch.js [hunk 2]
```diff
@@ -454,43 +454,51 @@ function startFetch(fetch, successcb, errorcb, progresscb, readystatechangecb) {
   var fetchAttrReplace = !!(fetchAttributes & {{{ cDefine('EMSCRIPTEN_FETCH_REPLACE') }}});
   var fetchAttrSynchronous = !!(fetchAttributes & {{{ cDefine('EMSCRIPTEN_FETCH_SYNCHRONOUS') }}});
 
+  function doCallback(f) {
+    if (fetchAttrSynchronous) {
+      f();
+    } else {
+      callUserCallback(f);
+    }
+  }
```
Comment (kripken) [inline src/Fetch.js:463]: Also depends on another PR?

**Comment by nokotan** [comment]
LGTM. The approach is different from #17376, #17376 can be closed.

File: src/Fetch.js (hunk 2 above)
Comment (sbc100) [reply src/Fetch.js]: Yes, this depends on #17440

Comment (sbc100) [reply src/Fetch.js]: Re-based onto main

File: src/library.js [hunk 3]
```diff
@@ -3506,6 +3506,10 @@ mergeInto(LibraryManager.library, {
       return;
     }
     try {
+#if ASSERTIONS
+      assert(!inUserCode, 'callUserCallback called when already running user code');
+      inUserCode = true;
+#endif
```
Comment (kripken) [inline src/library.js:3512]: Why is this an error?

Comment (sbc100) [reply src/library.js:3506]: If we did have such uses, then they would show up as test failures here, right?

If you are already in a user stack then you can just invoke the callback yourself.. what would be the point of using `callUserCallback` in that case?  You wouldn't want it to do any of the things it normally does... which is specifically why @dschuff added the, now-removed, `syncronous` argument.. which bypassed what `calllUserCallback` was actually doing.

Anyone assuming they can use `callUserCallback` when they are already in a user stack would already be broken, as in this `glfw` bug.

Comment (sbc100) [reply src/library.js:3506]: The first live of the docs seem pretty clear to me: `Used to call user callbacks from the embedder / event loop.`

If you are not coming from the event loop you simply don't need to call this function.

This new assertion just helps such users know that they are making a mistake.

Comment (sbc100) [reply src/library.js:3506]: How about if I made this into a warnOnce rather than an assertion?  I think that might be a better way to go.

**Comment by sbc100** [comment]
I reverted all the code to check for misused of callUserCallback.  We can leave futher discussion of that to #17476.

That should make this change a lot less controversial as it simply fixes a bug.

**Review by kripken — APPROVED** [review]

**Pull request merged**